key_value_store.h

```diff
@@ -0,0 +1,30 @@
+#pragma once
+#include <iostream>
+#include <unordered_map>
+#include <mutex>
+#include <boost/functional/hash.hpp>
+
+namespace std {
+template <typename T, typename A> struct hash<std::vector<T, A> > {
+  size_t operator()(const std::vector<T, A>& v) const {
+    return boost::hash_range(v.cbegin(), v.cend());
+  }
+};
+}
+
+struct FlaggedValue {
```

Comment:
Maybe provide a default constructor that initializes flags to 0? I don't think the compilers default constructor will do that for you.